File: 프로그래머스/lv4/64063. 호텔 방 배정/호텔 방 배정.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>

using namespace std;

map<long long, long long> hotel;
long long reservation(long long room) {
    auto itr = hotel.find(room);
    if(itr == hotel.end()) {
        hotel.insert({room, room+1});
        return room;
    } else {
        long long ll = reservation(itr->second);
        itr->second = ll;
        return ll;
    }
}

vector<long long> solution(long long k, vector<long long> room_number) {

    vector<long long> answer;
    for(long long room_now : room_number) {
        answer.emplace_back(reservation(room_now));
    }
    return answer;
}

File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.java
import java.util.*;

class Solution {
    
    Map<Long, Long> hotel = new HashMap<>();
    
    long reservation(long roomNumber) {
        // hotel
        if (!hotel.containsKey(roomNumber)) {
            hotel.put(roomNumber, roomNumber + 1);
            return roomNumber;
        } else {
            long room = reservation(hotel.get(roomNumber));
            hotel.put(roomNumber, room);
            return room;
        }
    }
    
    public long[] solution(long k, long[] roomNumbers) {
        long[] answer = new long[roomNumbers.length];
        for (int i = 0; i < roomNumbers.length; i++) {
            answer[i] = reservation(roomNumbers[i]);
        }
        return answer;
    }
}